--- /home/cdk/project/micofe/micofe/build/tmp/work/linux-6.12-r0/linux-6.12/soo/drivers/vinputback/vinput.c	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/drivers/vinputback/vinput.c	2026-05-01 11:12:17.081409871 +0200
@@ -0,0 +1,335 @@
+/*
+ * Copyright (C) 2026 Clément Dieperink <clement.dieperink@heig-vd.ch>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ */
+
+#if 0
+#define DEBUG
+#endif
+
+#include <linux/types.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/slab.h>
+#include <linux/delay.h>
+#include <linux/of.h>
+#include <linux/miscdevice.h>
+
+#include <uapi/linux/input.h>
+
+#include <soo/evtchn.h>
+#include <soo/gnttab.h>
+#include <soo/hypervisor.h>
+#include <soo/vbus.h>
+
+#include <soo/vdevback.h>
+
+#include <soo/dev/vinput.h>
+
+typedef struct {
+	/* Must be the first field */
+	vinput_t vinput;
+
+	/* Reference to the ring */
+	grant_ref_t ring_ref;
+
+} vinput_priv_t;
+
+/*
+ * We maintain a list of vdev_destination
+ */
+struct vdev_destination {
+	struct list_head list;
+	struct vbus_device *vdev_destination;
+};
+
+static struct miscdevice miscdev;
+static struct list_head vdev_destinations;
+
+
+/*
+ * Add a new destination for a new ME.
+ */
+static void add_destination(struct vbus_device *vdev_destination)
+{
+	struct vdev_destination *destination;
+
+	destination = kzalloc(sizeof(*destination), GFP_ATOMIC);
+	BUG_ON(!destination);
+
+	destination->vdev_destination = vdev_destination;
+
+	list_add(&destination->list, &vdev_destinations);
+}
+
+/*
+ * Search for a destination related to a specific ME according to its domid.
+ */
+static struct vbus_device *get_destination(uint32_t domid)
+{
+	struct vdev_destination *entry;
+
+	list_for_each_entry(entry, &vdev_destinations, list)
+		if (entry->vdev_destination->otherend_id == domid)
+			return entry->vdev_destination;
+	return NULL;
+}
+
+/*
+ * Remove a destination attached to a specific ME
+ */
+static void del_destination(struct vbus_device *vdev_destination)
+{
+	struct vdev_destination *entry;
+
+	list_for_each_entry(entry, &vdev_destinations, list)
+		if (entry->vdev_destination == vdev_destination) {
+			list_del(&entry->list);
+			kfree(entry);
+			return;
+		}
+	BUG();
+}
+
+static void process_request(struct vbus_device *vdev)
+{
+	vinput_priv_t *vinput_priv = dev_get_drvdata(&vdev->dev);
+	vinput_request_t *ring_req;
+
+	vdevback_processing_begin(vdev);
+
+	while ((ring_req = vinput_get_ring_request(&vinput_priv->vinput.ring)) !=
+	       NULL) {
+		/* Nothing to do ... */
+	}
+
+	vdevback_processing_end(vdev);
+}
+
+static irqreturn_t vinput_interrupt_bh(int irq, void *dev_id)
+{
+	struct vbus_device *vdev = (struct vbus_device *)dev_id;
+
+	process_request(vdev);
+
+	return IRQ_HANDLED;
+}
+
+static irqreturn_t vinput_interrupt(int irq, void *dev_id)
+{
+	/* All processing in the bottom half */
+
+	return IRQ_WAKE_THREAD;
+}
+
+/**
+ * This function is called in interrupt context.
+ * - If the state is Connected, the character can directly be pushed in the ring.
+ * - If the state is not Connected, the character is pushed into a circular buffer that
+ *   will be flushed at the next call to the Connected callback.
+ */
+void me_cons_send_input(domid_t domid, struct input_event *event)
+{
+	struct vbus_device *destination;
+	vinput_priv_t *vinput_priv;
+	vinput_response_t *vinput_rsp;
+
+	destination = get_destination(domid);
+
+	if (!destination || !vdevfront_is_connected(destination))
+		return;
+
+	vdevback_processing_begin(destination);
+
+	vinput_priv = dev_get_drvdata(&destination->dev);
+
+	spin_lock(&vinput_priv->vinput.ring_lock);
+	vinput_rsp = vinput_new_ring_response(&vinput_priv->vinput.ring);
+
+	vinput_rsp->type = event->type;
+	vinput_rsp->code = event->code;
+	vinput_rsp->value = event->value;
+
+	spin_unlock(&vinput_priv->vinput.ring_lock);
+
+	vinput_ring_response_ready(&vinput_priv->vinput.ring);
+
+	if (destination->state == VbusStateConnected)
+		notify_remote_via_virq(vinput_priv->vinput.irq);
+
+	vdevback_processing_end(destination);
+}
+
+static void vinput_probe(struct vbus_device *vdev)
+{
+	vinput_priv_t *vinput_priv;
+
+	vinput_priv = kzalloc(sizeof(vinput_priv_t), GFP_ATOMIC);
+	BUG_ON(!vinput_priv);
+
+	spin_lock_init(&vinput_priv->vinput.ring_lock);
+
+	dev_set_drvdata(&vdev->dev, vinput_priv);
+
+	add_destination(vdev);
+
+	DBG(VINPUT_PREFIX "Backend probe: %d\n", vdev->otherend_id);
+}
+
+static void vinput_remove(struct vbus_device *vdev)
+{
+	vinput_priv_t *vinput_priv = dev_get_drvdata(&vdev->dev);
+
+	del_destination(vdev);
+
+	DBG("%s: freeing the vinput structure for %s\n", __func__,
+	    vdev->nodename);
+	kfree(vinput_priv);
+}
+
+static void vinput_close(struct vbus_device *vdev)
+{
+	vinput_priv_t *vinput_priv = dev_get_drvdata(&vdev->dev);
+
+	DBG("(vinput) Backend close: %d\n", vdev->otherend_id);
+
+	/*
+	 * Free the ring and unbind evtchn.
+	 */
+
+	BACK_RING_INIT(&vinput_priv->vinput.ring, vinput_priv->vinput.ring.sring,
+		       PAGE_SIZE);
+	unbind_from_virqhandler(vinput_priv->vinput.irq, vdev);
+
+	gnttab_unmap(vinput_priv->vinput.ring.sring);
+
+	vinput_priv->vinput.ring.sring = NULL;
+}
+
+static void vinput_suspend(struct vbus_device *vdev)
+{
+	DBG("(vinput) Backend suspend: %d\n", vdev->otherend_id);
+}
+
+static void vinput_resume(struct vbus_device *vdev)
+{
+	DBG("(vinput) Backend resume: %d\n", vdev->otherend_id);
+}
+
+static void vinput_reconfigured(struct vbus_device *vdev)
+{
+	unsigned int evtchn;
+	vinput_sring_t *sring;
+	vinput_priv_t *vinput_priv = dev_get_drvdata(&vdev->dev);
+
+	DBG(VINPUT_PREFIX "Backend reconfigured: %d\n", vdev->otherend_id);
+
+	/*
+	 * Set up a ring (shared page & event channel) between the agency and the ME.
+	 */
+
+	vbus_gather(VBT_NIL, vdev->otherend, "ring-ref", "%lu",
+		    &vinput_priv->ring_ref, "ring-evtchn", "%u", &evtchn, NULL);
+
+	DBG("BE: ring-ref=%u, event-channel=%u\n", vinput_priv->ring_ref, evtchn);
+
+	gnttab_map(vdev->otherend_id, vinput_priv->ring_ref, (void **)&sring);
+
+	BACK_RING_INIT(&vinput_priv->vinput.ring, sring, PAGE_SIZE);
+
+	vinput_priv->vinput.irq = bind_interdomain_evtchn_to_virqhandler(
+		vdev->otherend_id, evtchn, vinput_interrupt, vinput_interrupt_bh,
+		0, VINPUT_NAME "-backend", vdev);
+}
+
+static void vinput_connected(struct vbus_device *vdev)
+{
+	DBG(VINPUT_PREFIX "Backend connected: %d\n", vdev->otherend_id);
+
+	process_request(vdev);
+}
+
+vdrvback_t vinputdrv = { .probe = vinput_probe,
+			.remove = vinput_remove,
+			.close = vinput_close,
+			.connected = vinput_connected,
+			.reconfigured = vinput_reconfigured,
+			.resume = vinput_resume,
+			.suspend = vinput_suspend };
+
+static ssize_t vinput_write(struct file *filp, const char __user *buf, size_t len, loff_t *ppos)
+{
+	size_t handled_event = 0;
+	size_t ret;
+	size_t event_count;
+	struct input_event __user *u_event;
+	struct input_event event;
+
+	if ((len % sizeof(struct input_event)) != 0) {
+		return -EINVAL;
+	}
+
+	u_event = (struct input_event __user *)buf;
+	event_count = len / sizeof(struct input_event);
+
+	for (handled_event = 0; handled_event < event_count; handled_event++) {
+		ret = copy_from_user(&event, &u_event[handled_event], sizeof(event));
+		if (ret != 0) {
+			return -EFAULT;
+		}
+
+		me_cons_send_input(vfb_get_focus(), &event);
+	}
+
+	return len;
+}
+
+static struct file_operations vinput_fops = {
+	.write = vinput_write,
+};
+
+static int vinput_init(void)
+{
+	struct device_node *np;
+	int ret;
+
+	np = of_find_compatible_node(NULL, NULL, "vinput,backend");
+
+	/* Check if DTS has vinput enabled */
+	if (!of_device_is_available(np))
+		return 0;
+
+	INIT_LIST_HEAD(&vdev_destinations);
+
+	vdevback_init(VINPUT_NAME, &vinputdrv);
+
+	memset(&miscdev, 0, sizeof(miscdev));
+	miscdev.minor = MISC_DYNAMIC_MINOR;
+	miscdev.name = VINPUT_NAME;
+	miscdev.mode = 0666;
+	miscdev.fops = &vinput_fops;
+
+	ret = misc_register(&miscdev);
+	if (ret != 0) {
+		printk(KERN_ERR VINPUT_PREFIX "Couldn't register misc device: %d\n", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+device_initcall(vinput_init);
