--- /home/rossierd/soo/micofe/27/build/tmp/work/linux-6.12-rpi4-r0/linux-6.12-rpi4/soo/drivers/Kconfig	1970-01-01 01:00:00.000000000 +0100
+++ ./soo/drivers/Kconfig	2026-03-13 13:24:35.605124126 +0100
@@ -0,0 +1,51 @@
+#
+# SOO device configuration
+#
+
+
+menu "Smart Object Oriented subsystem" 
+
+config VDUMMY_BACKEND
+        bool "SOO (realtime and no-RT) vDummy support for debugging and testing"
+
+config VUART_BACKEND
+	bool "vuart UART Backend support for SOO Mobile Entity consoles"
+	select SERIAL_CORE_CONSOLE
+	
+config VLEDS_BACKEND
+        bool "vleds LEDS backend driver"
+        default N
+        
+config VSENSELED_BACKEND
+	depends on FB_RPISENSE || ARCH_VEXPRESS
+	bool "vsenseled Sense-HAT led backend driver"
+	default n
+	help
+	  The vsenseled driver enables 5 logical LEDs that may be used to represent the joystick movement
+	  for example.
+
+config VSENSEJ_BACKEND
+	depends on FB_RPISENSE || ARCH_VEXPRESS
+	bool "vsensej Sense-HAT joystick backend driver"
+	default n
+	help
+	  The vsensej driver captures joystick movements.
+
+config VLOGS_BACKEND
+	bool "vlogs LOGS Backend support for SOO Capsules"
+	help
+	  The vlogs driver is used to logs Capsules logs - logs are store in
+	  `/var/log/soo/capsule_N.log` (N corresponds to the SlotID of the capsule).
+
+config VFBDEV_BACKEND
+	bool "vfbdev Framebuffer Backend support for SOO Capsules"
+	help
+	  The vfbdev driver is used to allow capsules to access a framebuffer from Linux.
+
+config VINPUT_BACKEND
+	bool "vinput Input support for SOO Capsules"
+	help
+	  The vinput driver is used to allow capsules to access a mouse/keyboard inputs from Linux.
+
+endmenu
+
